### PR TITLE
Core: Add total data size to Partitions table

### PR DESCRIPTION
**Pull request by hsiang-c**
Closes https://github.com/apache/iceberg/issues/7896

This PR adds `total_data_file_size_in_bytes` to Partitions Table

### CODE REVIEW

File: docs/flink-queries.md [hunk 1]
```diff
@@ -436,7 +436,7 @@ SELECT * FROM prod.db.table$partitions;
 | {20211002, 10} | 1            | 1          | 0       |
 
 Note:
-For unpartitioned tables, the partitions table will contain only the record_count and file_count columns.
+For unpartitioned tables, the partitions table will contain only the record_count, file_count and total_data_size_in_bytes columns.
```
Comment (ajantha-bhat) [inline docs/flink-queries.md:439]: This statement needs to be updated. Unpartitioned table will have few more fields and previous PR missed to update this. 

https://github.com/apache/iceberg/blob/4e6c7bad35ae3886378ec50a79dac0368df875a8/core/src/main/java/org/apache/iceberg/PartitionsTable.java#L96-L105

File: docs/spark-queries.md [hunk 2]
```diff
@@ -319,7 +319,7 @@ SELECT * FROM prod.db.table.partitions;
 |  {20211002, 10}|           1|         1|         0|
 
 Note:
-1. For unpartitioned tables, the partitions table will contain only the record_count and file_count columns.
+1. For unpartitioned tables, the partitions table will contain only the record_count, file_count and total_data_size_in_bytes columns.
```
Comment (ajantha-bhat) [inline docs/spark-queries.md:322]: This statement needs to be updated. Unpartitioned table will have few more fields and previous PR missed to update this. 

https://github.com/apache/iceberg/blob/4e6c7bad35ae3886378ec50a79dac0368df875a8/core/src/main/java/org/apache/iceberg/PartitionsTable.java#L96-L105

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java [hunk 3]
```diff
@@ -82,7 +82,12 @@ public class PartitionsTable extends BaseMetadataTable {
                 10,
                 "last_updated_snapshot_id",
                 Types.LongType.get(),
-                "Id of snapshot that last updated this partition"));
+                "Id of snapshot that last updated this partition"),
+            Types.NestedField.required(
+                11,
+                "total_data_size_in_bytes",
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/iceberg/PartitionsTable.java:88]: For partition stats, I am using the field name as `data_file_size_in_bytes` here.

https://github.com/apache/iceberg/pull/7105/files#diff-36347a47c3bf67ea2ef6309ea96201814032d21bb5f162dfae4045508c15588aR736

Let us wait and see what others think about the naming of this field. Based on that either you or me have to modify accordingly.

**Comment by ajantha-bhat** [comment]
cc: @szehon-ho as you are mostly working and reviewing this area.

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 3 above)
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java:88]: Ah good point, I'm ok with total_data_file_size_in_bytes then to be consistent, initially thought its a bit long.

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java [hunk 4]
```diff
@@ -82,7 +82,12 @@ public class PartitionsTable extends BaseMetadataTable {
                 10,
                 "last_updated_snapshot_id",
                 Types.LongType.get(),
-                "Id of snapshot that last updated this partition"));
+                "Id of snapshot that last updated this partition"),
+            Types.NestedField.required(
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/PartitionsTable.java:86]: And this topic always comes up, but what do you think of the position?  @ajantha-bhat @dramaticlly .  Maybe its better after file_count?  (so we have 3 columns for data, pos_delete, and eq_delete)

**Review by szehon-ho — COMMENTED** [review]
Thanks , yea I was chatting earlier with @hsiang-c for this :)

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 4 above)
Comment (dramaticlly) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java:86]: Yeah I think what Szehon said make sense, given last 2 columns are optional and new are required

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 5]
```diff
@@ -1290,6 +1295,12 @@ public void testUnpartitionedPartitionsTable() {
             .set("position_delete_file_count", 0)
             .set("equality_delete_record_count", 0L)
             .set("equality_delete_file_count", 0)
+            .set(
+                "total_data_size_in_bytes",
+                StreamSupport.stream(
+                        table.currentSnapshot().addedDataFiles(table.io()).spliterator(), false)
+                    .mapToLong(DataFile::fileSizeInBytes)
+                    .sum())
```
Comment (dramaticlly) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1303]: probably worth extract a variable instead of inline the computation. 

also I saw you added coverage for unpartitioned table only, shall we also add one for partitioned table to make sure it s data size in bytes match for each partition?

Comment (hsiang-c) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1303]: @dramaticlly Thank you for your feedback. 

Yes, we should add tests for partitioned table. I was able to do it for `testPartitionsTable` but not `testPartitionsTableDeleteStats` and `testPartitionsTableLastUpdatedSnapshot`.

Will dig into it more today.

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 3 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java:88]: I am also ok with `total_data_file_size_in_bytes`

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 4 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java:86]: Agree. I have already kept it beside `file_count` for partition stats. 

Note: Here we should not modify field id while reordering to maintain the compatibility.

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java [hunk 6]
```diff
@@ -73,6 +73,11 @@ public class PartitionsTable extends BaseMetadataTable {
                 "equality_delete_file_count",
                 Types.IntegerType.get(),
                 "Count of equality delete files"),
+            Types.NestedField.required(
+                11,
+                "total_data_file_size_in_bytes",
+                Types.LongType.get(),
+                "Total bytes of data files in a partition"),
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/PartitionsTable.java:80]: nit: 'total size in bytes'

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 7]
```diff
@@ -2028,4 +2042,10 @@ public static Dataset<Row> selectNonDerived(Dataset<Row> metadataTable) {
   public static Types.StructType nonDerivedSchema(Dataset<Row> metadataTable) {
     return SparkSchemaUtil.convert(selectNonDerived(metadataTable).schema()).asStruct();
   }
+
+  private long getDataFileSizeInBytes(Iterable<DataFile> dataFiles) {
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:2046]: nit: we can remove 'get' (Iceberg code style guideline are a bit different:  https://iceberg.apache.org/contribute/#method-naming)

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java [hunk 8]
```diff
@@ -275,6 +283,7 @@ static class Partition {
     private int eqDeleteFileCount;
     private Long lastUpdatedMs;
     private Long lastUpdatedSnapshotId;
+    private long dataFileSizeInBytes;
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/PartitionsTable.java:286]: nit: can we move after dataFileCount?  (as its part of 'dataFile' group)

**Review by szehon-ho — COMMENTED** [review]
Looks good, just some style nits

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java [hunk 9]
```diff
@@ -73,6 +73,11 @@ public class PartitionsTable extends BaseMetadataTable {
                 "equality_delete_file_count",
                 Types.IntegerType.get(),
                 "Count of equality delete files"),
+            Types.NestedField.required(
+                11,
+                "total_data_file_size_in_bytes",
+                Types.LongType.get(),
+                "Total size in bytes"),
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/PartitionsTable.java:80]: Ah sorry, in my previous comment I meant just change "total bytes' => 'total size in bytes', but the rest was ok.

So can we revert back the original end of sentence?  

'Total size in bytes of data files'  (maybe 'in a partition' was redundant there)

File: docs/flink-queries.md (hunk 1 above)
Comment (hsiang-c) [reply docs/flink-queries.md]: @ajantha-bhat Good catch, fixed.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java (hunk 5 above)
Comment (hsiang-c) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java]: @dramaticlly I think I fixed `testPartitionsTableLastUpdatedSnapshot`, please take a look, thanks!

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 8 above)
Comment (dramaticlly) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java]: speak of which, @szehon-ho do you feel we shall do the same in Schema method to move this new field with id 11 to be right after file_count (field id 3)? It seem to fit into same dataFile group by it might be some concern about reference by position to mess up?

File: docs/flink-queries.md [hunk 10]
```diff
@@ -436,7 +436,7 @@ SELECT * FROM prod.db.table$partitions;
 | {20211002, 10} | 1            | 1          | 0       |
 
 Note:
-For unpartitioned tables, the partitions table will contain only the record_count and file_count columns.
+For unpartitioned tables, the partitions table will contain only the record_count, file_count, position_delete_record_count, position_delete_file_count, equality_delete_record_count, equality_delete_file_count, last_updated_ms, last_updated_snapshot_id and total_data_file_size_in_bytes columns.
```
Comment (szehon-ho) [inline docs/flink-queries.md:439]: Should we do this in another pr?  I feel we need to edit the table above as well.

Also, I think we can just say 'For unpartitioned tables, the partitions table will not contain the partition and spec_id field', as the list of fields we do support is becoming too big.

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java [hunk 11]
```diff
@@ -73,6 +73,8 @@ public class PartitionsTable extends BaseMetadataTable {
                 "equality_delete_file_count",
                 Types.IntegerType.get(),
                 "Count of equality delete files"),
+            Types.NestedField.required(
+                11, "total_data_file_size_in_bytes", Types.LongType.get(), "Total size in bytes"),
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/PartitionsTable.java:77]: This is still not changed back?  "Total size in bytes of data files"  Sorry if its still pending

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 8 above)
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java]: Oh yea , i think that was the consensus from the other comemnt: https://github.com/apache/iceberg/pull/7920#discussion_r1244089947  @hsiang-c do you think we can move it?

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 11 above)
Comment (dramaticlly) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java:77]: and also let's move it up to between 3 and 5 since it belong to data file group

File: docs/flink-queries.md (hunk 10 above)
Comment (dramaticlly) [reply docs/flink-queries.md:439]: agreed, we can follow up with doc PR after this is merged

File: core/src/main/java/org/apache/iceberg/PartitionsTable.java (hunk 8 above)
Comment (hsiang-c) [reply core/src/main/java/org/apache/iceberg/PartitionsTable.java]: @szehon-ho Sure thing.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 12]
```diff
@@ -1421,6 +1435,7 @@ public void testPartitionsTable() {
     }
   }
 
+  @SuppressWarnings("MethodLength")
```
Comment (hsiang-c) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1438]: `testPartitionsTableLastUpdatedSnapshot` has more than 150 lines due to the new assertions I added.  Please let me know if they make sense to you, thanks.

Comment (szehon-ho) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1438]: I think we can have some helper methods.  I'm ok to do it in a follow up pr, maybe @dramaticlly can also help as the original author ?

Comment (dramaticlly) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1438]: yeah I also run into this problem before. If your datafile total size assertion is already covered in `testPartitionsTable` then I dont think it's necessary to ido it again in `testPartitionsTableLastUpdatedSnapshot` since what you tested are mostly the same scenario. 

Also if you want to move file assertion into some helper method to reduce line count, that's probably more preferred than suppress warning. 

I can also help with follow up PR if needed, it's up to you

Comment (szehon-ho) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1438]: I was thinking we can make a helper for making the expected?  But not sure if we can do it so that  spotless wont blow it back up to one field per line

Comment (szehon-ho) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1438]: Also some other things like, example:

```
    Assert.assertEquals(
        "rewrite replaced 2 manifests",
        2,
        Iterables.size(rewriteManifestResult.rewrittenManifests()));
    Assert.assertEquals(
        "rewrite added 1 manifests", 1, Iterables.size(rewriteManifestResult.addedManifests()));
```

can be

```
assertResult(rwriteManifestResult, 2, 1)

Comment (hsiang-c) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java]: @dramaticlly @szehon-ho Thanks for your feedback.

Extracted those assertions to a helper method, see https://github.com/apache/iceberg/pull/7920/commits/0d423c219739a1e26d167d7a2bd176a2cb516544

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 13]
```diff
@@ -2223,4 +2254,36 @@ private DeleteFile writeEqDeleteFile(Table table) {
...
+  }
+
+  private List<DataFile> listDataFilesFromCommitId(Table table, long commitId) {
+    return StreamSupport.stream(
+            table.snapshot(commitId).addedDataFiles(table.io()).spliterator(), false)
+        .collect(Collectors.toList());
+  }
+
+  private void assertPartitionIdFromDataFiles(
+      List<DataFile> dataFilesFromFirstCommit, List<DataFile> dataFilesFromSecondCommit) {
```
Comment (dramaticlly) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:2271]: I am thinking if we write the helper method this way, then it probably cannot be reused. I would suggest to put assertion expected value as parameters (like numberOfDataFiles and partitionValue), everything else looks good to me

Comment (hsiang-c) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java]: Good point! fixed now.

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 14]
```diff
@@ -1469,12 +1483,20 @@ public void testPartitionsTableLastUpdatedSnapshot() {
         new GenericRecordBuilder(
             AvroSchemaUtil.convert(
                 partitionsTable.schema().findType("partition").asStructType(), "partition"));
+
+    List<DataFile> dataFilesFromFirstCommit = listDataFilesFromCommitId(table, firstCommitId);
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1487]: Would it work to make a method List<DataFile> dataFiles(table) to get all the data files, so we don't have to do add data files from both commits?

I did this before here: https://github.com/apache/iceberg/blob/master/spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewritePositionDeleteFilesAction.java#L682

(I think we can do it without column stats here).

If we do this, we can even extract to TestHelpers in a later PR.

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 15]
```diff
@@ -2028,4 +2055,32 @@ public static Dataset<Row> selectNonDerived(Dataset<Row> metadataTable) {
...
   }
+
+  private long sumDataFileSizeInBytes(Iterable<DataFile> dataFiles) {
+    return StreamSupport.stream(dataFiles.spliterator(), false)
+        .mapToLong(DataFile::fileSizeInBytes)
+        .sum();
+  }
+
+  private List<DataFile> listDataFilesFromCommitId(Table table, long commitId) {
+    return StreamSupport.stream(
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:2066]: Or we can not have this method and instead use https://github.com/apache/iceberg/blob/master/spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewritePositionDeleteFilesAction.java#L682 as in the other comment.

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java (hunk 14 above)
Comment (hsiang-c) [reply spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1487]: @szehon-ho Didn't know this approach, thanks!

> If we do this, we can even extract to TestHelpers in a later PR.
+1, let's do the extraction in a later PR.

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 16]
```diff
@@ -1469,12 +1483,20 @@ public void testPartitionsTableLastUpdatedSnapshot() {
         new GenericRecordBuilder(
             AvroSchemaUtil.convert(
                 partitionsTable.schema().findType("partition").asStructType(), "partition"));
+
+    List<DataFile> dataFilesFromFirstCommit = listDataFilesFromCommitId(table, firstCommitId);
+    assertPartitionIdFromDataFiles(dataFilesFromFirstCommit, 2, Arrays.asList(1, 2));
```
[the remark this answers is not in the document]
Comment (hsiang-c) [reply spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1488]: @szehon-ho Do you mean `new int[] {1, 2}` or `List.of(1, 2)`?

[the remark this answers is not in the document]
Comment (hsiang-c) [reply spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java]: Fixed!

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 17]
```diff
@@ -2028,4 +2053,22 @@ public static Dataset<Row> selectNonDerived(Dataset<Row> metadataTable) {
...
+  private long totalSizeInBytes(Iterable<DataFile> dataFiles) {
+    return Lists.newArrayList(dataFiles).stream().mapToLong(DataFile::fileSizeInBytes).sum();
+  }
+
+  private List<DataFile> dataFiles(Table table) {
+    CloseableIterable<FileScanTask> tasks = table.newScan().planFiles();
+    return Lists.newArrayList(CloseableIterable.transform(tasks, FileScanTask::file));
+  }
+
+  private void assertDataFilePartitions(List<DataFile> dataFiles, int[] expectedPartitionIds) {
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:2066]: Nit: we can put back the size check.

File: spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 18]
```diff
@@ -1469,12 +1484,18 @@ public void testPartitionsTableLastUpdatedSnapshot() {
         new GenericRecordBuilder(
             AvroSchemaUtil.convert(
                 partitionsTable.schema().findType("partition").asStructType(), "partition"));
+
+    List<DataFile> dataFiles = dataFiles(table);
+    Assert.assertEquals("Table should have 3 data files", 3, dataFiles.size());
+    assertDataFilePartitions(dataFiles, new int[] {1, 2, 2});
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:1490]: Sorry I'm trying to remember why I said the previous comment, I was probably thinking varargs, like 
`assertDataFilePartitions(List<DataFile>, int... partitions)` so then it would just be ` assertDataFilePartitions(dataFiles, 1,2,3)`, but on second look it may not be so clean.
 
I think Arrays.asList() is equally long, let's revert to that, and it may make the length check easier inside the method?

**Review by szehon-ho — APPROVED** [review]

**Comment by szehon-ho** [comment]
Merged , thanks a lot @hsiang-c for the first contribution, and thanks @ajantha-bhat and @dramaticlly for additional reviews!

**Comment by szehon-ho** [comment]
(Feel free to make follow prs to update the docs)
